sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -514,6 +514,19 @@ case class OptimizeCodegen(conf: CatalystConf) extends Rule[LogicalPlan] {
 
 
 /**
+ * Reorders the predicates in `Filter` so more expensive expressions like UDF can evaluate later.
+ */
+object ReorderPredicatesInFilter extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case Filter(pred, child) =>
+      val (deterministicExprs, others) = splitConjunctivePredicates(pred).reverse
```

Comment:
Yea I see. I only reorder the deterministic expression in the suffix of predicates.